app/jobs/bulkrax/create_relationships_job.rb

```diff
@@ -165,13 +172,34 @@ def add_to_collection(child_record, parent_record)
     end
 
     def add_to_work(child_record, parent_record)
-      return true if parent_record.ordered_members.to_a.include?(child_record)
+      parent_record.is_a?(Valkyrie::Resource) ? add_to_valkyrie_work(child_record, parent_record) : add_to_af_work(child_record, parent_record)
 
-      parent_record.ordered_members << child_record
       @parent_record_members_added = true
       @child_members_added << child_record
     end
 
+    def add_to_valkyrie_work(child_record, parent_record)
+      return true if parent_record.member_ids.include?(child_record.id)
+
+      parent_record.member_ids << child_record.id
+
+      # TODO: Hyrax is in the process of extracting an "Action" object that we could call.  It does
+      # provide validation that we may want to consider.
+      #
+      # NOTE: We may need to look at the step args we're passing, see
+      #      `Hyrax::WorksControllerBehavior#update_valkyrie_work`
+      #      Hyrax's `./app/controllers/concerns/hyrax/works_controller_behavior.rb`
+      #
+      change_set = Hyrax::ChangeSet.for(parent_record)
+      Hyrax::Transactions::Container['change_set.update_work'].call(change_set)
```

Comment:
i'm curious about this choice. won't this ensure that the entire update work process (the same one that happens via the controller action) takes place just to change the work membership?

Hyrax::Publisher provides an `object.membership.updated` event, and i wonder whether it wouldn't be better to save the work with the updated membership and publish that? https://github.com/samvera/hyrax/blob/fdcabe651850b1885d7bd3e1b36777bf5f336ee8/lib/hyrax/publisher.rb#L183-L192

Reply:
@no-reply, for our uses, we're trying to hit this [transaction](https://github.com/scientist-softserv/iiif_print/blob/9dec8418c879cf5f9346e4105ae221bdd9946902/app/transactions/hyrax/transactions/steps/set_child_flag.rb), so you're suggesting we put a publisher here and then that means this can become a listener?

Reply:
[here](https://github.com/scientist-softserv/iiif_print/blob/b4694c1b33cba86161768ea7f1bd9fecbf1b083c/app/listeners/iiif_print/listener.rb) was the original idea of the listener

Reply:
@kirkkwang yes. i think what i'm suggesting (and i'm just popping in as an outsider here), is that you want this behavior to trigger whenever an object's membership changes, not only when a given transaction runs.

making it a `listener` means you'll trigger these changes when outside callers  (in Hyrax or in the downstream application) change membership, as long as they dutifully publish the expected message.

i think a helpful line to draw around transactions and listeners is:
  - use a transaction step when you want to EXPLICITLY implement some behavior as part of a given operation, but you might want to (explicitly) implement different behavior in a different context.
    - e.g. i might want to set the `current_user` as the "depositor" every time a `create` action happens in the hyrax works controller, but handle this matter differently if i create an object through an API endpoint.
    - it would be normal for the Hyrax engine, Bulkrax's own code, or the downstream application to call different transactions that may or may not include steps the others might use.
   - use a listener when you want to ensure some behavior *always* runs in response to some event, regardless of that event's context.
     - e.g. whenever metadata updates for an object, i want to reindex that object.
 
another way of thinking about it is:
  - if in a naive MVC architecture i would have just written the code in the controller, or as a model method, like `do_this; do_that; do_other`, then you want that to be a transaction step.
  -  if in a naive MVC architecture i would have written a callback, you want a listener. (it might be helpful to think of a listener as a multiplexed callback, decoupled from the model layer).

Reply:
Thanks @no-reply that's super helpful!

Reply:
FYI I made a ticket here https://github.com/scientist-softserv/iiif_print/issues/330

Reply:
@no-reply should [this](https://github.com/samvera/hyrax/blob/double_combo/lib/hyrax/transactions/steps/add_to_parent.rb#L19-L32) also publish `object.membership.updated`?  I see that it there's a membership update but it publishes `object.metadata.updated`.  Can/should it do both?